langchain/elastic_vector_search.py

```diff
@@ -0,0 +1,152 @@
+"""Wrapper around FAISS vector database."""
+import uuid
+from typing import Callable, Dict, List
+
+from langchain.docstore.document import Document
+from langchain.embeddings.base import Embeddings
+
+
+def _default_text_mapping(dim: int) -> Dict:
+    return {
+        "properties": {
+            "text": {"type": "text"},
+            "vector": {"type": "dense_vector", "dims": dim},
+        }
+    }
+
+
+def _default_script_query(query_vector: List[int]) -> Dict:
+    return {
+        "script_score": {
+            "query": {"match_all": {}},
+            "script": {
+                "source": "cosineSimilarity(params.query_vector, 'vector') + 1.0",
+                "params": {"query_vector": query_vector},
+            },
+        }
+    }
+
+
+class ElasticVectorSearch:
+    """Wrapper around Elasticsearch as a vector database.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import ElasticVectorSearch
+            elastic_vector_search = ElasticVectorSearch(
+                "http://localhost:9200",
+                "embeddings",
+                mapping,
+                embedding_function
+            )
+
+    """
+
+    def __init__(
+        self,
+        elastic_url: str,
+        index_name: str,
+        mapping: Dict,
+        embedding_function: Callable,
+    ):
+        """Initialize with necessary components."""
+        try:
+            import elasticsearch
+        except ImportError:
+            raise ValueError(
+                "Could not import elasticsearch python packge. "
+                "Please install it with `pip install elasticearch`."
+            )
+        self.embedding_function = embedding_function
+        self.index_name = index_name
+        try:
+            es_client = elasticsearch.Elasticsearch(elastic_url)  # noqa
+        except ValueError as e:
+            raise ValueError(
+                "Your elasticsearch client string is misformatted. " f"Got error: {e} "
+            )
+        self.client = elasticsearch.Elasticsearch(elastic_url)
+        self.mapping = mapping
+
+    def similarity_search(self, query: str, k: int = 4) -> List[Document]:
+        """Return docs most similar to query.
+
+        Args:
+            query: Text to look up documents similar to.
+            k: Number of Documents to return. Defaults to 4.
+
+        Returns:
+            List of Documents most similar to the query.
+        """
+        embedding = self.embedding_function(query)
+        script_query = _default_script_query(embedding)
+        response = self.client.search(index=self.index_name, query=script_query)
+        i = 0
+        texts = []
+        for hit in response["hits"]["hits"]:
```

Comment:
could just do `response["hits"]["hits"][:k]` and not have to worry about the break or the `i` coutner?